baggage_restriction_manager.go

```diff
@@ -0,0 +1,153 @@
+	IsValidBaggageKey(key string) (bool, int32)
+}
+
+// DefaultBaggageRestrictionManager allows any baggage key.
+type DefaultBaggageRestrictionManager struct{}
+
+// IsValidBaggageKey implements BaggageRestrictionManager#IsValidBaggageKey
+func (m DefaultBaggageRestrictionManager) IsValidBaggageKey(key string) (bool, int32) {
+	return true, defaultMaxValueLength
+}
+
+type httpBaggageRestrictionsManager struct {
+	serverURL string
+}
+
+func (s *httpBaggageRestrictionsManager) GetBaggageRestrictions(serviceName string) ([]*baggage.BaggageRestriction, error) {
+	var out []*baggage.BaggageRestriction
+	v := url.Values{}
+	v.Set("service", serviceName)
+	if err := utils.GetJSON(s.serverURL+"?"+v.Encode(), &out); err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+type baggageRestrictionManager struct {
+	baggageRestrictionManagerOptions
+	sync.RWMutex
+
+	serviceName     string
+	restrictionsMap map[string]int32
+	timer           *time.Ticker
+	thriftManager   baggage.BaggageRestrictionManager
+	pollStopped     sync.WaitGroup
+	stopPoll        chan struct{}
+}
+
+// NewBaggageRestrictionManager returns a BaggageRestrictionManager that polls the agent for the latest
+// baggage restrictions.
+func NewBaggageRestrictionManager(
+	serviceName string,
+	options ...BaggageRestrictionManagerOption,
+) (BaggageRestrictionManager, error) {
+	opts := applyOptions(options...)
+	m := &baggageRestrictionManager{
+		serviceName:                      serviceName,
+		baggageRestrictionManagerOptions: opts,
+		restrictionsMap:                  make(map[string]int32),
+		timer:                            time.NewTicker(opts.refreshInterval),
+		thriftManager:                    &httpBaggageRestrictionsManager{serverURL: opts.baggageRestrictionManagerServerURL},
+		stopPoll:                         make(chan struct{}),
+	}
+	if err := m.updateRestrictions(); err != nil {
+		return nil, fmt.Errorf("Failed to initialize baggage restrictions: %s", err.Error())
+	}
+	m.pollStopped.Add(1)
+	go m.pollManager()
+	return m, nil
+}
+
+func (m *baggageRestrictionManager) IsValidBaggageKey(key string) (bool, int32) {
+	m.RLock()
+	defer m.RUnlock()
+	if maxValueLength, ok := m.restrictionsMap[key]; ok {
+		return true, maxValueLength
+	}
+	return false, 0
+}
+
+// Close implements Close() of Sampler.
+func (m *baggageRestrictionManager) Close() {
+	close(m.stopPoll)
+	m.pollStopped.Wait()
+}
+
+func (m *baggageRestrictionManager) pollManager() {
+	defer m.pollStopped.Done()
+	defer m.timer.Stop()
+
+	for {
+		select {
+		case <-m.timer.C:
+			if err := m.updateRestrictions(); err != nil {
+				m.logger.Error(fmt.Sprintf("Failed to update baggage restrictions: %s", err.Error()))
+			}
+		case <-m.stopPoll:
+			return
+		}
+	}
+}
+
+func (m *baggageRestrictionManager) updateRestrictions() error {
+	restrictions, err := m.thriftManager.GetBaggageRestrictions(m.serviceName)
+	if err != nil {
+		// TODO this is pretty serious, almost retry worthy
```

Comment:
at 1 minute intervals, I was thinking of being more aggressive on failure (ie try 3 more times), maybe over zealous